### PR TITLE
Add Cinder Max Volume Limit

### PR DESCRIPTION
**Pull request by gnufied**
We currently have 3 specific predicates that filter and count attachable limits for AWS, Azure and GCE-PD. This PR brings Cinder in-line with these other volume types.  

This PR also implements initial support for reporting volume limits on per-node basis for Openstack nodes. The implementation here however provides a placeholder value which can be overridden.

```release-note
Add support for max attach limit for Cinder
```

### CODE REVIEW

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 1]
```diff
@@ -370,6 +377,8 @@ func getMaxVolumeFunc(filterName string) func(node *v1.Node) int {
 			return DefaultMaxGCEPDVolumes
 		case AzureDiskVolumeFilterType:
 			return DefaultMaxAzureDiskVolumes
+		case CinderVolumeFilterType:
+			return volumeutil.DefaultMaxCinderVolumes
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Unrelated: Why do we have `DefaultMaxCinderVolumes` coming from volumeutil, whereas rest(like DefaultMaxAzureDiskVolumes etc) are coming from the current file

Comment (gnufied) [reply pkg/scheduler/algorithm/predicates/predicates.go]: I think it is an oversight. All of these defaults should come from volumeutil. I will move them over.

Comment (gnufied) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Or more specifically - `volumeutil.attach_limit.go` was introduced as common lib file that could be shared between plugin and scheduler. Most plugins use this file to derive keys and default limits. But scheduler already had those constants and they weren't removed and replaced when this new file was created.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -558,6 +567,23 @@ var AzureDiskVolumeFilter = VolumeFilter{
 	},
 }
 
+// CinderVolumeFilter is a VolumeFilter for filtering Cinder Volumes
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Please add some comments on when you want to remove these cloud provider specific code and rely on CSI plugin.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 1 above)
Comment (ravisantoshgudimetla) [reply pkg/scheduler/algorithm/predicates/predicates.go]: ok.

**Comment by gnufied** [comment]
cc @dims

**Comment by dims** [comment]
cc @flaper87 @adisky @hogepodge @j-griffith

**Comment by dims** [comment]
@gnufied i expected [BlockStorageOpts](https://cs.k8s.io/?q=BlockStorageOpts&i=nope&files=&repos=kubernetes/kubernetes) would be updated to allow an admin to configure the value (default of 256 is ok!)

**Comment by childsb** [comment]
this is a serious issue for people using cinder.  approved from sig-storage
/approve

**Comment by ravisantoshgudimetla** [comment]
@gnufied - Seems there is go-lint issue here, https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/72980/pull-kubernetes-verify/119121. Can you fix it as well?

**Review by j-griffith — APPROVED** [review]
/lgtm seems worthwhile to get folks using the Cinder provider up to speed with the other cloud providers that already added this predicate.

**Comment by ravisantoshgudimetla** [comment]
Tracking issue to remove cloud provider specific providers in scheduler.

https://github.com/kubernetes/kubernetes/issues/72920


@gnufied - Assuming that the above issue will be handled and all the cloud provider specific predicates could be replaced with CSI plugin, I am fine with changes from scheduler side.

/approve

/cc @bsalamat - fyi

**Comment by bsalamat** [comment]
I am not really fine with this change. The largest cloud providers today have limits which should be removed. Urgency of a customer issue or other similar reasons are not good justifications to worsen the quality of our code.

**Comment by dims** [comment]
@bsalamat i had requested a specific change as well and it definitely seems to have been merged in a rush for some reason! i'd support a revert and then a proper follow up

**Comment by ravisantoshgudimetla** [comment]
>  The largest cloud providers today have limits which should be removed.

@bsalamat - I understand your concern. As discussed offline, I am going to create a PR which mentions explicitly that the predicates that are cloud provider specific are deprecated in this release and in subsequent releases, we plan to remove the entire cloud provider specific predicates and replace them with csi predicate.

> i had requested a specific change as well and it definitely seems to have been merged in a rush for some reason! i'd support a revert and then a proper follow up

@dims - I thought you're fine with a follow-up PR and gave my approval. To be clear, @bsalamat's concern here is adding code for cinder where as IIUC your concern is more on the lines of limits to be coming from cloudprovider's config file. So, do you still want this to be handled within scheduler code considering we are going to deprecate it?

**Comment by dims** [comment]
@ravisantoshgudimetla Merging a PR within 3 hours in a rush with pending comments does not look good from an external perception. especially with all the lgtm-ers and approve-ers from one company. Also, we should try to conduct business in public channels as well. i don't think i mentioned that i was ok with a follow up PR (but that would have been a matter of record if we were not doing things off-line).

**Comment by gnufied** [comment]
@dims I hadn't intended this PR to get merged fast, I was still thinking about how best to address your concerns when lgtms and approvals rolled in and the bot took over. Sorry about it. 

We are going to talk about this PR in sig-storage community meeting. @dims I will follow up with your concerns - https://github.com/kubernetes/kubernetes/issues/72988

**Comment by dims** [comment]
Thanks @gnufied

**Comment by spiffxp** [comment]
FWIW I am a big fan of using `/hold` to tell our automation that a PR is not yet ready for merge for reasons that can't be expressed otherwise

**Comment by childsb** [comment]
My review & direct discussion regarding the PR had a couple identified issues being addressed in a follow up PR.  Next time I'll make sure to use a hold so that everyone has time to review any direct communications.  Also this is both a bug/feature, so we were moving fast to try and fix the bug aspect.  Anyways, we should have held the PR- apologies @dims

**Comment by dims** [comment]
Thanks @childsb I see the follow up PR for my concern from @gnufied in #72988. We should file a follow up issue at least for the predicates as well.

**Comment by resouer** [comment]
Emm ... as Dims commented above, I don't think  #72988 reflect the concern from sig-scheduling , @bsalamat , @dims  and other folks in this thread. cc @gnufied @childsb 

I believe the concern is about we are not tend to add any cloud provider related predicates for reason of customer needs, in which context, I would +1 for revert this change for now cc @dims . 

The original issue could be addressed in follow up discussion with sig-scheduling + sig-storage.

**Comment by gnufied** [comment]
@dims @resouer @bsalamat So to avoid any confusion - can you guys please give us your final opinion on this? Should we revert it or not?

@saad-ali @childsb lets finalize this from sig-storage perspective too.

**Comment by dims** [comment]
@gnufied Most of the code in this PR is in `pkg/scheduler` so i'll defer to sig-scheduling folks / @bsalamat for making the call. 

cc @kubernetes/sig-scheduling-pr-reviews @k82cn

**Comment by derekwaynecarr** [comment]
I do not think this should be reverted.  The PR makes no impact to quality, it brings a feature up to same quality as the existing public cloud providers.  I see no reason why we are advantaging public cloud providers.

**Comment by derekwaynecarr** [comment]
To be clear, I agree with the proposal outlined here:
https://github.com/kubernetes/kubernetes/pull/72980#issuecomment-455299811

**Comment by derekwaynecarr** [comment]
I agree that the PR should have been open longer for more comment prior to merge.  Apologies for that issue as noted above, but still believe the project should ensure users are broadly successful.

**Comment by yastij** [comment]
I do agree here with @bsalamat - having providers predicates already there is not a reason to add new ones (same as cloudprovider and storage providers).

cc @kubernetes/sig-cloud-provider for the extraction effort

**Comment by bsalamat** [comment]
I support removing all cloud provider specific code for a more generic approach. I do think that this PR has lowered the quality of our code base. The larger such "exceptions" are, the harder it will be to vote against adding more of them in the future. So, existence of three of them does not justify adding the fourth one.

All being said, I am fine with leaving this PR merged as long as there is a plan to remove all of the similar cloud provider specific code from our code base. I spoke with @ravisantoshgudimetla offline and he has agreed to come up with a plan for 1.14 to remove the cloud provider specific code.

**Comment by ravisantoshgudimetla** [comment]
Thanks @bsalamat, the plan @gnufied and I agreed upon is to deprecate the hardcoded cloud provider specific predicates in 1.14 so that people will have time to start using CSI based one and in subsequent releases, we will remove the predicates.

**Comment by resouer** [comment]
I am fine as #72920 is raised.

**Comment by dims** [comment]
whew! thanks everyone for the discussion, decision and future path. as someone who has been championing OpenStack/Kubernetes integration i am pleased.

**Pull request merged**